test/builder/pipeline_test.go

```diff
@@ -121,10 +131,14 @@ func TestPipelineRun(t *testing.T) {
 			PipelineRef:     v1alpha1.PipelineRef{Name: "tomatoes"},
 			ServiceAccount:  "sa",
 			ServiceAccounts: []v1alpha1.PipelineRunSpecServiceAccount{{TaskName: "foo", ServiceAccount: "sa-2"}},
-			Params: []v1alpha1.Param{{
-				Name:  "first-param",
-				Value: "first-value",
-			}},
+			Params: []v1alpha1.ArrayOrStringParam{
```

Comment:
Are we going to have to change the type of this `Params` field if we decide in the future we want to support `ArrayOrStringOrIntOrBoolOrStructParam`? :)

If so, maybe this indicates we should come up with a type name that doesn't tie us to the types available within it. I don't suppose we could reuse `Param`...?

Reply:
Retyping this comment because it looks like it never sent when I typed it 2 days ago:

Currently `Param` is an existing type used exclusively by `PipelineResources` and only can hold a string (used for things like a git url). I definitely like the idea of refactoring `ArrayOrStringParam`, though, because it is a more general parameter type than the current `Param` type is. 

I pushed a code change earlier that renames the existing `Param` type to `ResourceParam`, which I think is more specific and appropriate, and and renames `ArrayOrStringParam` to `Param`. 

Let me know if that change is good; one thing to note is that for future changes, the `Param` type name will never change, but its `Value` field, which is currently of type `ArrayOrString` would have to change to `ArrayOrStringOrIntOrBoolOrStruct`. I don't think that's avoidable.

Reply:
Blech. I think we should try to figure out if there's some other (backward-compatible!) way to handle this should we decide we need to in future changes, but I don't want to block this PR on solving that.

The rest of that change lgtm.